modules/geochemistry/src/userobjects/NodalVoidVolume.C

```diff
@@ -0,0 +1,226 @@
+
+  // Build _nodes_to_send using global_node_nums_to_send, keeping the same ordering in the
+  // std::vector
+  _nodes_to_send.clear();
+  for (const auto & kv : global_node_nums_to_send)
+  {
+    const processor_id_type pid = kv.first;
+    _nodes_to_send[pid] = std::vector<const Node *>();
+    for (const auto & node_num : kv.second)
+      _nodes_to_send[pid].push_back(_mesh.nodePtr(node_num));
+  }
+}
+
+void
+NodalVoidVolume::exchangeGhostedInfo()
+{
+  // Exchange ghosted _nodal_void_volume information with other processors
+  std::map<processor_id_type, std::vector<Real>> nvv_to_send;
+  for (const auto & kv : _nodes_to_send)
+  {
+    const processor_id_type pid = kv.first;
+    nvv_to_send[pid] = std::vector<Real>();
+    for (const auto & nd : kv.second)
+      nvv_to_send[pid].push_back(_nodal_void_volume.at(nd));
+  }
+
+  auto nvv_action_functor = [this](processor_id_type pid, const std::vector<Real> & nvv_received) {
+    const std::size_t msg_size = nvv_received.size();
+    mooseAssert(msg_size == _nodes_to_receive[pid].size(),
+                "Message size, " << msg_size
+                                 << ", incompatible with nodes_to_receive, which has size "
+                                 << _nodes_to_receive[pid].size());
+    for (unsigned i = 0; i < msg_size; ++i)
+      _nodal_void_volume[_nodes_to_receive[pid][i]] =
+          _nodal_void_volume.at(_nodes_to_receive[pid][i]) + nvv_received[i];
+  };
+  Parallel::push_parallel_vector_data(this->comm(), nvv_to_send, nvv_action_functor);
+}
+
+void
+NodalVoidVolume::initialize()
+{
+  for (auto & nvv : _nodal_void_volume)
+    nvv.second = 0.0;
+}
+
+void
+NodalVoidVolume::finalize()
+{
+  if (_app.n_processors() > 1)
+    exchangeGhostedInfo();
+  // Now _nodal_void_volume is correct for all nodes within and on the boundary of this processor's
+  // domain.
+}
+
+void
+NodalVoidVolume::threadJoin(const UserObject & uo)
+{
+  // _nodal_void_volume will have been computed by other threads: add their contributions to ours.
+  const NodalVoidVolume & nvv = static_cast<const NodalVoidVolume &>(uo);
+  for (auto & our_nvv : _nodal_void_volume)
+    our_nvv.second += nvv._nodal_void_volume.at(our_nvv.first);
+  // Now _nodal_void_volume is correct for all nodes within this processor's domain (but potentially
+  // not those on the boundary: exchangeGhostedInfo() will fix that)
+}
+
+void
+NodalVoidVolume::execute()
+{
+  // this gets called for all elements owned by this processor.  So, after threadJoin(), it
+  // correctly computes _nodal_void_volume for nodes compltely within this processor's domain.
+  for (unsigned i = 0; i < _current_elem->n_nodes(); ++i)
+  {
+    for (unsigned qp = 0; qp < _qrule->n_points(); ++qp)
+      _nodal_void_volume[_current_elem->node_ptr(i)] =
+          _nodal_void_volume.at(_current_elem->node_ptr(i)) +
+          _JxW[qp] * _coord[qp] * _phi[i][qp] * _porosity[qp];
+  }
```

Comment:
```suggestion
  for (const auto & node : _current_elem->node_ref_range())
  {
    auto & node_entry = _nodal_void_volume[&node];
    for (unsigned qp = 0; qp < _qrule->n_points(); ++qp)
      entry += _JxW[qp] * _coord[qp] * _phi[i][qp] * _porosity[qp];
  }
```
Here you're looking into the map a staggering `2 * _qrule->n_points()` times! This makes that only once.

Reply:
Thanks - good optimisation here!